pkg/ruler/remotequerier.go

```diff
@@ -309,11 +310,25 @@ func (q *RemoteQuerier) sendRequest(ctx context.Context, req *httpgrpc.HTTPReque
 			}
 			return resp, nil
 		}
-		// 4xx errors shouldn't be retried because it is expected that
-		// running the same query gives rise to the same 4xx error.
-		if code := grpcutil.ErrorToStatusCode(err); code/100 == 4 {
-			return nil, err
+
+		// Bail out if the error is known to be not retriable.
+		switch code := grpcutil.ErrorToStatusCode(err); code {
+		case codes.ResourceExhausted:
```

Comment:
As discussed via Slack, the initial idea was to make all errors with `codes.ResourceExhausted` status code fail. Some of these errors come from gRPC itself (e.g., https://github.com/grpc/grpc-go/blob/84b85babc00b4b8460e53b6ee110bfb49e9311cf/rpc_util.go#L610-L615). On the other hand, there are some errors with status code `codes.ResourceExhausted` returned by distributors and ingesters that should be retried. These cases are:
1. ingester too busy: the returned error has `codes.ResourceExhausted` status code, but also contains a `mimirpb.ErrorDetail` with cause `mimirpb.TOO_BUSY` and should **always** be retried
2. ingestion rate limit hit: the returned error has `codes.ResourceExhausted` status code, but also contains a `mimirpb.ErrorDetail` with cause `mimirpb.INGESTION_RATE_LIMITED` and should be retried if distributor's `-distributor.service-overload-status-code-on-rate-limit-enabled` is set to true
3. request rate limit hit: the returned error has `codes.ResourceExhausted` status code, but also contains a `mimirpb.ErrorDetail` with cause `mimirpb.REQUEST_RATE_LIMITED` and should be retried if distributor's `-distributor.service-overload-status-code-on-rate-limit-enabled` is set to true

So I recommend to check whether the error contains a detail of type `mimirpb.ErrorDetails` with one of the causes mentioned above and to retry when needed.

An example:
```
stat, ok := grpcutil.ErrorToStatus(err)
if ok {
	switch code := stat.Code() {
		case codes.ResourceExhausted:
			details := stat.Details()
			if len(details) != 1 {
				return nil, err
			}
			errDetails, ok := details[0].(*mimirpb.ErrorDetails)
			if !ok {
				return nil, err
			}
			if errDetails.GetCause() == mimirpb.INGESTION_RATE_LIMITED 
				|| errDetails.GetCause() == mimirpb.REQUEST_RATE_LIMITED
				|| errDetails.GetCause() == mimirpb.TOO_BUSY {
				continue
			}
			return nil, err
		default:
			// In case the error was a wrapped HTTPResponse, its code represents HTTP status;
			// 4xx errors shouldn't be retried because it is expected that
			// running the same query gives rise to the same 4xx error.
			if code/100 == 4 {
				return nil, err
			}
		}
}
```
IMPORTANT: this code is just an example, and it doesn't take into account the `-distributor.service-overload-status-code-on-rate-limit-enabled` flag.

Reply:
After another chat with @narqo I actually realized that `mimirpb.INGESTION_RATE_LIMITED` and `mimirpb.REQUEST_RATE_LIMITED` could be thrown on the write path only, while the present PR concerns the read path. Therefore, the `-distributor.service-overload-status-code-on-rate-limit-enabled` CLI flag is not important, and the implementation can be simplified.

```
stat, ok := grpcutil.ErrorToStatus(err)
if ok {
	switch code := stat.Code() {
		case codes.ResourceExhausted:
			details := stat.Details()
			if len(details) != 1 {
				return nil, err
			}
			errDetails, ok := details[0].(*mimirpb.ErrorDetails)
			if !ok {
				return nil, err
			}
			if errDetails.GetCause() == mimirpb.TOO_BUSY {
				continue
			}
			return nil, err
		default:
			// In case the error was a wrapped HTTPResponse, its code represents HTTP status;
			// 4xx errors shouldn't be retried because it is expected that
			// running the same query gives rise to the same 4xx error.
			if code/100 == 4 {
				return nil, err
			}
		}
}
```